xbmc/platform/linux/input/LibInputHandler.cpp

```diff
@@ -0,0 +1,249 @@
+    return;
+  }
+
+  libinput_log_set_handler(m_li, LogHandler);
+  libinput_log_set_priority(m_li, LIBINPUT_LOG_PRIORITY_DEBUG);
+
+  auto ret = libinput_udev_assign_seat(m_li, "seat0");
+  if (ret < 0)
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - failed to assign seat", __FUNCTION__);
+
+  m_liFd = libinput_get_fd(m_li);
+
+  m_keyboard.reset(new CLibInputKeyboard(winEvents));
+  m_pointer.reset(new CLibInputPointer(winEvents));
+  m_touch.reset(new CLibInputTouch());
+}
+
+CLibInputHandler::~CLibInputHandler()
+{
+  if (m_li)
+    libinput_unref(m_li);
+
+  if (m_udev)
+    udev_unref(m_udev);
+}
+
+void CLibInputHandler::OnReadyRead()
+{
+  auto ret = libinput_dispatch(m_li);
+  if (ret < 0)
+  {
+    CLog::Log(LOGERROR, "CLibInputHandler::%s - libinput_dispatch failed", __FUNCTION__);
+    return;
+  }
+
+  libinput_event *ev;
+  while ((ev = libinput_get_event(m_li)) != nullptr)
+  {
+    ProcessEvent(ev);
+    libinput_event_destroy(ev);
+  }
+}
+
+void CLibInputHandler::ProcessEvent(libinput_event *ev)
+{
+  libinput_event_type type = libinput_event_get_type(ev);
+  libinput_device *dev = libinput_event_get_device(ev);
+
+  switch (type)
+  {
+    case LIBINPUT_EVENT_DEVICE_ADDED:
+    {
+      const char *sysname = libinput_device_get_sysname(dev);
+      const char *name = libinput_device_get_name(dev);
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_TOUCH))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device == m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - touch type device added: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.push_back(dev);
+        }
+      }
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_POINTER))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device == m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - pointer type device added: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.push_back(dev);
+        }
+      }
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_KEYBOARD))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device == m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - keyboard type device added: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.push_back(dev);
+        }
+        m_keyboard->GetRepeat(dev);
+      }
+      break;
+    }
+    case LIBINPUT_EVENT_DEVICE_REMOVED:
+    {
+      const char *sysname = libinput_device_get_sysname(dev);
+      const char *name = libinput_device_get_name(dev);
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_TOUCH))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device != m_devices.end())
```

Comment:
see the ADDED comment, this should always be true. and you need to `libinput_device_unref()`